docs/internals/optimizer.rst

```diff
@@ -0,0 +1,1251 @@
+.. index:: optimizer, optimiser, common subexpression elimination, constant propagation
+.. _optimizer:
+
+*************
+The Optimizer
+*************
+
+The Solidity compiler uses two different optimizer modules: The "old" optimizer
+that operates at opcode level and the "new" optimizer that operates on Yul IR code.
+
+The opcode-based optimizer applies a set of `simplification rules <https://github.com/ethereum/solidity/blob/develop/libevmasm/RuleList.h>`_
+to opcodes. It also combines equal code sets and removes unused code.
+
+The Yul-based optimizer is much more powerful, because it can work across function
+calls: In Yul, it is not possible to perform arbitrary jumps, so it is for example
+possible to compute the side-effects of each function. Consider two function calls,
+where the first does not modify the storage and the second modifies the storage. If their arguments and return values does not depend on each other, we can reorder the function calls. Similarly, if a function is
+side-effect free and its result is multiplied by zero, you can remove the function
+call completely.
+
+Currently, the parameter ``--optimize`` activates the opcode-based optimizer for the
+generated bytecode and the Yul optimizer for the Yul code generated internally, for example for ABI coder v2.
+One can use ``solc --ir-optimized --optimize`` to produce
+optimized experimental Yul IR for a Solidity source. Similarly, use ``solc --strict-assembly --optimize``
+for a stand-alone Yul mode.
+
+You can find more details on both optimizer modules and their optimization steps below.
+
+Benefits of Optimizing Solidity Code
+====================================
+
+Overall, the optimizer tries to simplify complicated expressions, which reduces both code
+size and execution cost. It also specializes or inlines functions. Especially
```

Comment:
```suggestion
size and execution cost, i.e., it can reduce gas needed for contract deployment as well as external calls to the contract. It also specializes or inlines functions. Especially
```
It's probably good to include the word `gas` somewhere.